mixnet/queuesim/simulation.py

```diff
@@ -0,0 +1,142 @@
+from framework.usim import Framework
+from protocol.nomssip import NomssipConfig
+from protocol.temporalmix import TemporalMixConfig, TemporalMixType
+from queuesim.config import Config
+from queuesim.node import Node
+from sim.config import LatencyConfig
+from sim.connection import MeteredRemoteSimplexConnection
+from sim.topology import build_full_random_topology
+
+
+class Simulation:
+    """
+    Manages the entire cycle of simulation: initialization, running, and analysis.
+    """
+
+    def __init__(self, config: Config):
+        self.config = config
+
+    async def run(self):
+        async with usim.Scope() as scope:
+            self.framework = Framework(scope)
+            self.message_builder = MessageBuilder(self.framework)
+            self.dissemination_times = await self.__run()
+            self.framework.stop_tasks()
+
+    async def __run(self) -> list[float]:
+        self.received_msg_queue: Queue[tuple[float, bytes]] = self.framework.queue()
+
+        # Run and connect nodes
+        nodes = self.__run_nodes()
+        self.__connect_nodes(nodes)
+
+        # Choose and start senders
+        senders = self.config.sender_generator.choices(nodes, k=self.config.num_senders)
+        for sender in senders:
+            self.framework.spawn(self.__run_sender(sender))
+
+        # To count how many nodes have received each message
+        received_msg_counters: Counter[bytes] = Counter()
+        # To collect the dissemination times of each message.
+        dissemination_times: list[float] = []
+        # Wait until all messages are disseminated to the entire network.
+        while len(dissemination_times) < self.config.num_sent_msgs:
+            # Wait until a node notifies that it has received a new message.
+            received_time, msg = await self.received_msg_queue.get()
+            # If the message has been received by all nodes, calculate the dissemination time.
+            received_msg_counters.update([msg])
+            if received_msg_counters[msg] == len(nodes):
+                dissemination_times.append(
+                    received_time - Message.from_bytes(msg).sent_time
+                )
+        return dissemination_times
+
+    def __run_nodes(self) -> list[Node]:
+        return [
+            Node(
+                self.framework,
+                self.config.nomssip,
+                self.__process_msg,
+            )
+            for _ in range(self.config.num_nodes)
+        ]
+
+    async def __process_msg(self, msg: bytes) -> None:
+        """
+        A handler to process messages received via Nomos Gossip channel
+        """
+        # Notify that a new message has been received by the node.
+        # The received time is also included in the notification.
+        await self.received_msg_queue.put((self.framework.now(), msg))
+
+    def __connect_nodes(self, nodes: list[Node]):
+        topology = build_full_random_topology(
+            rng=self.config.topology.seed,
```

Comment:
A nit: shouldn't we use `seed` instead of `rng` here? For me `rng` is an algorithm...

Reply:
`rng` is a better name. But, I named it as `seed` because it's the field name in the config.yaml, and the seed value is automatically converted into `random.Random` by the `dacite` library (used to load a yaml file into an object).
I've also notices this name mismatch, but couldn't have a neat solution yet. I will find a better way soon.

(Of course, we can let the Config object have a seed (integer) and we can build a `random.Random` object from the `seed` when we want. But, I wanted to make the RNG as a singleton in the Config object, to avoid potential bugs.)

Reply:
Great, no need to push here if the current way is more convenient then I'm fine with that.